Resource cleanup sometimes times out

Given the next action after resource cleanup often would be to destroy the whole deployment and continue with the next bundle I would argue we should just log the fact and carry on. Alternatively we should change the values making up the retry behavior.
```
2020-09-15 15:34:16 [INFO] Removing instances launched by test (zaza-octavia*)
2020-09-15 15:36:42 [INFO] FAIL
2020-09-15 15:36:42 [INFO] ======================================================================
2020-09-15 15:36:42 [INFO] FAIL: test_create_loadbalancer (zaza.openstack.charm_tests.octavia.tests.LBAASv2Test)
2020-09-15 15:36:42 [INFO] Create load balancer.
2020-09-15 15:36:42 [INFO] ----------------------------------------------------------------------
2020-09-15 15:36:42 [INFO] Traceback (most recent call last):
2020-09-15 15:36:42 [INFO]   File "/tmp/tmp.QUz9aiXKju/func/lib/python3.5/site-packages/zaza/openstack/charm_tests/test_utils.py", line 119, in tearDown
2020-09-15 15:36:42 [INFO]     self.resource_cleanup()
2020-09-15 15:36:42 [INFO]   File "/tmp/tmp.QUz9aiXKju/func/lib/python3.5/site-packages/zaza/openstack/charm_tests/octavia/tests.py", line 97, in resource_cleanup
2020-09-15 15:36:42 [INFO]     super(LBAASv2Test, self).resource_cleanup()
2020-09-15 15:36:42 [INFO]   File "/tmp/tmp.QUz9aiXKju/func/lib/python3.5/site-packages/zaza/openstack/charm_tests/test_utils.py", line 497, in resource_cleanup
2020-09-15 15:36:42 [INFO]     msg="server")
2020-09-15 15:36:42 [INFO]   File "/tmp/tmp.QUz9aiXKju/func/lib/python3.5/site-packages/zaza/openstack/utilities/openstack.py", line 2014, in delete_resource
2020-09-15 15:36:42 [INFO]     resource_removed(resource, resource_id, msg)
2020-09-15 15:36:42 [INFO]   File "/tmp/tmp.QUz9aiXKju/func/lib/python3.5/site-packages/zaza/openstack/utilities/openstack.py", line 1995, in resource_removed
2020-09-15 15:36:42 [INFO]     msg)
2020-09-15 15:36:42 [INFO]   File "/tmp/tmp.QUz9aiXKju/func/lib/python3.5/site-packages/tenacity/__init__.py", line 409, in call
2020-09-15 15:36:42 [INFO]     do = self.iter(retry_state=retry_state)
2020-09-15 15:36:42 [INFO]   File "/tmp/tmp.QUz9aiXKju/func/lib/python3.5/site-packages/tenacity/__init__.py", line 368, in iter
2020-09-15 15:36:42 [INFO]     raise retry_exc.reraise()
2020-09-15 15:36:42 [INFO]   File "/tmp/tmp.QUz9aiXKju/func/lib/python3.5/site-packages/tenacity/__init__.py", line 186, in reraise
2020-09-15 15:36:42 [INFO]     raise self.last_attempt.result()
2020-09-15 15:36:42 [INFO]   File "/usr/lib/python3.5/concurrent/futures/_base.py", line 398, in result
2020-09-15 15:36:42 [INFO]     return self.__get_result()
2020-09-15 15:36:42 [INFO]   File "/usr/lib/python3.5/concurrent/futures/_base.py", line 357, in __get_result
2020-09-15 15:36:42 [INFO]     raise self._exception
2020-09-15 15:36:42 [INFO]   File "/tmp/tmp.QUz9aiXKju/func/lib/python3.5/site-packages/tenacity/__init__.py", line 412, in call
2020-09-15 15:36:42 [INFO]     result = fn(*args, **kwargs)
2020-09-15 15:36:42 [INFO]   File "/tmp/tmp.QUz9aiXKju/func/lib/python3.5/site-packages/zaza/openstack/utilities/openstack.py", line 1958, in _resource_removed
2020-09-15 15:36:42 [INFO]     assert len(matching) == 0, "Resource {} still present".format(resource_id)
2020-09-15 15:36:42 [INFO] AssertionError: Resource 55ca0e70-b61a-409d-bf1e-7ed2833319c1 still present
```

--- a/zaza/openstack/charm_tests/octavia/diskimage_retrofit/setup.py
+++ b/zaza/openstack/charm_tests/octavia/diskimage_retrofit/setup.py
@@ -15,6 +15,7 @@
 """Code for configuring octavia-diskimage-retrofit."""
 
 import logging
+import tenacity
 
 import zaza.model
 
@@ -39,12 +40,20 @@ def retrofit_amphora_image(unit='octavia-diskimage-retrofit/0',
     if image_id:
         params.update({'source-image': image_id})
 
-    # NOTE(fnordahl) ``zaza.model.run_action_on_leader`` fails here,
-    # apparently has to do with handling of subordinates in ``libjuju`` or
-    # ``juju`` itself.
-    action = zaza.model.run_action(
-        unit,
-        'retrofit-image',
-        action_params=params,
-        raise_on_failure=True)
+    # NOTE(fnordahl) the retrofit process involves downloading packages from
+    # the internet and is as such susceptible to random failures due to
+    # internet gremlins.
+    for attempt in tenacity.Retrying(
+            stop=tenacity.stop_after_attempt(3),
+            wait=tenacity.wait_exponential(
+            multiplier=1, min=2, max=10)):
+        with attempt:
+            # NOTE(fnordahl) ``zaza.model.run_action_on_leader`` fails here,
+            # apparently has to do with handling of subordinates in ``libjuju``
+            # or ``juju`` itself.
+            action = zaza.model.run_action(
+                unit,
+                'retrofit-image',
+                action_params=params,
+                raise_on_failure=True)
     return action

--- a/zaza/openstack/charm_tests/octavia/tests.py
+++ b/zaza/openstack/charm_tests/octavia/tests.py
@@ -18,6 +18,7 @@ import logging
 import subprocess
 import tenacity
 
+from keystoneauth1 import exceptions as keystone_exceptions
 import osc_lib.exceptions
 
 import zaza.openstack.charm_tests.test_utils as test_utils
@@ -80,8 +81,15 @@ class LBAASv2Test(test_utils.OpenStackBaseTest):
         # List of floating IPs created by this test
         cls.fips = []
 
-    def resource_cleanup(self):
-        """Remove resources created during test execution."""
+    @tenacity.retry(stop=tenacity.stop_after_attempt(3),
+                    wait=tenacity.wait_exponential(
+                        multiplier=1, min=2, max=10))
+    def resource_cleanup(self, only_local=False):
+        """Remove resources created during test execution.
+
+        :param only_local: When set to true do not call parent method
+        :type only_local: bool
+        """
         for lb in self.loadbalancers:
             self.octavia_client.load_balancer_delete(lb['id'], cascade=True)
             try:
@@ -90,8 +98,16 @@ class LBAASv2Test(test_utils.OpenStackBaseTest):
                     provisioning_status='DELETED')
             except osc_lib.exceptions.NotFound:
                 pass
+        # allow resource cleanup to be run multiple times
+        self.loadbalancers = []
         for fip in self.fips:
             self.neutron_client.delete_floatingip(fip)
+        # allow resource cleanup to be run multiple times
+        self.fips = []
+
+        if only_local:
+            return
+
         # we run the parent resource_cleanup last as it will remove instances
         # referenced as members in the above cleaned up load balancers
         super(LBAASv2Test, self).resource_cleanup()
@@ -157,6 +173,7 @@ class LBAASv2Test(test_utils.OpenStackBaseTest):
                     'provider': provider,
                 }})
         lb = result['loadbalancer']
+        self.loadbalancers.append(lb)
         lb_id = lb['id']
 
         logging.info('Awaiting loadbalancer to reach provisioning_status '
@@ -283,10 +300,25 @@ class LBAASv2Test(test_utils.OpenStackBaseTest):
         for provider in self.get_lb_providers(self.octavia_client).keys():
             logging.info('Creating loadbalancer with provider {}'
                          .format(provider))
-            lb = self._create_lb_resources(self.octavia_client, provider,
-                                           vip_subnet_id, subnet_id,
-                                           payload_ips)
-            self.loadbalancers.append(lb)
+            final_exc = None
+            # NOTE: we cannot use tenacity here as the method we call into
+            # already uses it to wait for operations to complete.
+            for retry in range(0, 3):
+                try:
+                    lb = self._create_lb_resources(self.octavia_client,
+                                                   provider,
+                                                   vip_subnet_id,
+                                                   subnet_id,
+                                                   payload_ips)
+                    break
+                except (AssertionError,
+                        keystone_exceptions.connection.ConnectFailure) as e:
+                    logging.info('Retrying load balancer creation, last '
+                                 'failure: "{}"'.format(str(e)))
+                    self.resource_cleanup(only_local=True)
+                    final_exc = e
+            else:
+                raise final_exc
 
             lb_fp = openstack_utils.create_floating_ip(
                 self.neutron_client, 'ext_net', port={'id': lb['vip_port_id']})

--- a/zaza/openstack/charm_tests/octavia/tests.py
+++ b/zaza/openstack/charm_tests/octavia/tests.py
@@ -19,6 +19,7 @@ import subprocess
 import tenacity
 
 from keystoneauth1 import exceptions as keystone_exceptions
+import octaviaclient.api.v2.octavia
 import osc_lib.exceptions
 
 import zaza.openstack.charm_tests.test_utils as test_utils
@@ -81,6 +82,27 @@ class LBAASv2Test(test_utils.OpenStackBaseTest):
         # List of floating IPs created by this test
         cls.fips = []
 
+    def _remove_amphorae_instances(self):
+        """Remove amphorae instances forcefully.
+
+        In some situations Octavia is unable to remove load balancer resources.
+        This helper can be used to remove the underlying instances.
+        """
+        result = self.octavia_client.amphora_list()
+        for amphora in result.get('amphorae', []):
+            for server in self.nova_client.servers.list():
+                if 'compute_id' in amphora and server.id == amphora[
+                        'compute_id']:
+                    try:
+                        openstack_utils.delete_resource(
+                            self.nova_client.servers,
+                            server.id,
+                            msg="server")
+                    except AssertionError as e:
+                        logging.warning(
+                            'Gave up waiting for resource cleanup: "{}"'
+                            .format(str(e)))
+
     @tenacity.retry(stop=tenacity.stop_after_attempt(3),
                     wait=tenacity.wait_exponential(
                         multiplier=1, min=2, max=10))
@@ -91,13 +113,21 @@ class LBAASv2Test(test_utils.OpenStackBaseTest):
         :type only_local: bool
         """
         for lb in self.loadbalancers:
-            self.octavia_client.load_balancer_delete(lb['id'], cascade=True)
             try:
-                self.wait_for_lb_resource(
-                    self.octavia_client.load_balancer_show, lb['id'],
-                    provisioning_status='DELETED')
-            except osc_lib.exceptions.NotFound:
-                pass
+                self.octavia_client.load_balancer_delete(
+                    lb['id'], cascade=True)
+            except octaviaclient.api.v2.octavia.OctaviaClientException as e:
+                logging.info('Octavia is unable to delete load balancer: "{}"'
+                             .format(e))
+                logging.info('Attempting to forcefully remove amphorae')
+                self._remove_amphorae_instances()
+            else:
+                try:
+                    self.wait_for_lb_resource(
+                        self.octavia_client.load_balancer_show, lb['id'],
+                        provisioning_status='DELETED')
+                except osc_lib.exceptions.NotFound:
+                    pass
         # allow resource cleanup to be run multiple times
         self.loadbalancers = []
         for fip in self.fips:

--- a/zaza/openstack/charm_tests/test_utils.py
+++ b/zaza/openstack/charm_tests/test_utils.py
@@ -495,6 +495,11 @@ class OpenStackBaseTest(BaseCharmTest):
                         self.nova_client.servers,
                         server.id,
                         msg="server")
+        except AssertionError as e:
+            # Resource failed to be removed within the expected time frame,
+            # log this fact and carry on.
+            logging.warning('Gave up waiting for resource cleanup: "{}"'
+                            .format(str(e)))
         except AttributeError:
             # Test did not define self.RESOURCE_PREFIX, ignore.
             pass

--- a/zaza/openstack/utilities/openstack.py
+++ b/zaza/openstack/utilities/openstack.py
@@ -690,14 +690,20 @@ def add_interface_to_netplan(server_name, mac_address):
                   "{}\nserver_name: {}".format(body_value, unit_name,
                                                interface, mac_address,
                                                server_name))
-    with tempfile.NamedTemporaryFile(mode="w") as netplan_file:
-        netplan_file.write(body_value)
-        netplan_file.flush()
-        model.scp_to_unit(unit_name, netplan_file.name,
-                          '/home/ubuntu/60-dataport.yaml', user="ubuntu")
-    run_cmd_mv = "sudo mv /home/ubuntu/60-dataport.yaml /etc/netplan/"
-    model.run_on_unit(unit_name, run_cmd_mv)
-    model.run_on_unit(unit_name, "sudo netplan apply")
+    for attempt in tenacity.Retrying(
+            stop=tenacity.stop_after_attempt(3),
+            wait=tenacity.wait_exponential(
+            multiplier=1, min=2, max=10)):
+        with attempt:
+            with tempfile.NamedTemporaryFile(mode="w") as netplan_file:
+                netplan_file.write(body_value)
+                netplan_file.flush()
+                model.scp_to_unit(
+                    unit_name, netplan_file.name,
+                    '/home/ubuntu/60-dataport.yaml', user="ubuntu")
+            run_cmd_mv = "sudo mv /home/ubuntu/60-dataport.yaml /etc/netplan/"
+            model.run_on_unit(unit_name, run_cmd_mv)
+            model.run_on_unit(unit_name, "sudo netplan apply")
 
 
 def configure_gateway_ext_port(novaclient, neutronclient, net_id=None,

